utils/retrieve-state/src/lib.rs

```diff
@@ -289,50 +337,369 @@ pub fn put_block_with_deploys(
-            progress_fn()
-        } else {
-            return Err(anyhow::anyhow!(
-                "unable to download trie at {:?}",
-                next_trie_key
-            ));
+    println!("got {}/{} valid peers", peer_map.len(), peers.len());
+
+    let mut total_tries_count = 0;
+
+    let mut outstanding_trie_keys = vec![state_root_hash];
+    let mut tries_downloaded = 0usize;
+    let mut in_flight_counters = peers
+        .iter()
+        .map(|addr| (addr, 0usize))
+        .collect::<HashMap<_, _>>();
+
+    loop {
+        // Top level task keeps track of missing descendants and work distribution.
+        for (address, peer) in peer_map.iter() {
+            let counter = in_flight_counters.get_mut(&address).unwrap();
+            // TODO: understand what's happening when requests get backed up, causing starvation of
+            // base_recv
+            if *counter < concurrent_requests_per_peer {
+                if let Some(next_trie_key) = outstanding_trie_keys.pop() {
+                    let mut sender = peer.sender.clone();
+                    *counter += 1;
+                    sender.send(PeerMsg::GetTrie(next_trie_key)).await.unwrap();
+                } else {
+                    break;
+                }
+            }
+        }
+
+        if let Some(msg) = base_recv.next().await {
+            match msg {
+                Ok(PeerMsg::TrieDownloaded {
+                    trie,
+                    peer,
+                    elapsed: _,
+                }) => {
+                    let engine_state = Arc::clone(&engine_state);
+                    /*
+                    let mut missing_trie_descendants = tokio::task::spawn_blocking(move || {
+                        engine_state
+                            .put_trie_and_find_missing_descendant_trie_keys(
+                                CorrelationId::new(),
+                                &trie,
+                            )
+                            .unwrap()
+                    })
+                    .await
+                    .unwrap();
+                    */
+                    let mut missing_trie_descendants = engine_state
+                        .put_trie_and_find_missing_descendant_trie_keys(CorrelationId::new(), &trie)
+                        .unwrap();
+
+                    // count of all tries we know about
+                    total_tries_count += missing_trie_descendants.len();
+
+                    outstanding_trie_keys.append(&mut missing_trie_descendants);
+
+                    // in-flight requests
+                    if let Some(counter) = in_flight_counters.get_mut(&peer) {
+                        *counter = counter.saturating_sub(1);
+                    }
+
+                    // count of all tries downloaded
+                    tries_downloaded += 1;
+
+                    if tries_downloaded % 1000 == 0 {
+                        println!(
+                            "tries downloaded: {downloaded}/{total}",
```

Comment:
Tries or MB / sec might be helpful here...

Reply:
Also, this is stalling for me... and it would be nice to have timestamps.

Maybe switch use logging like we have in the node/execution engine?